### ISSUE TITLE
renaming of metadata folder for development branch

### ISSUE THREAD

**Issue by vsoch** [apptainer/singularity#528]
This is my mistake - but @satra suggested (and thinking it through I agree) that we should not have the meta data folder be hidden. Instead, something along the lines of this:

```bash
/singularity.d
  /environment
	10-docker.sh
	99-userdef.sh
	..etc
  labels.json
```

@gmkurtzer thoughts?

**Comment by remyd1** [comment]
Probably that

```
/singularity.conf.d
  00-labels.json
  10-docker.sh
  99-userdef.sh
```

or with:

`/singularity.cond.d -> /etc/singularity.conf.d`

would be more appropriate.

Rémy

**Comment by satra** [comment]
@remyd1 `cond` -> `conf`

**Comment by wpoely86** [comment]
Fully agree that it should not be hidden.

**Comment by vsoch** [comment]
@gmkurtzer just making sure that you see this!

**Comment by gmkurtzer** [comment]
I'm not opposed to removing the leading ".", but I'd like to hear the reasoning.

On the other hand, I don't like the suffixes proposed as they don't truly represent or describe the content of the directory.

Thanks!

**Comment by vsoch** [comment]
hmm so you don't think `.d` is appropriate for a set of configuration files?

http://askubuntu.com/questions/7648/many-directories-have-a-d-suffix-extension-what-does-it-mean

it's really more for a "header and metadata stuffs" place until we have a better idea, haha. It also has runscript and execution things, so hmm. 

Anyway, I did like `singularity.d` I think for most that don't have linux backgrounds (like myself) I just thought it was a signal of "directory" and "there is config stuff in here" I do agree it's nice to not have it hidden, so the explorers among the group will see it and go "aha! this is relevant to my singularity image!"

@gmkurtzer what are your thoughts about other naming conventions?

**Comment by gmkurtzer** [comment]
Typically directories that include a suffix of `.d` have been used to describe files of all the same format and in a flat directory structure (e.g. `/etc/httpd/conf.d` and `/etc/rc.d/init.d`/). The `/.singularity` directory is not flat, nor does it contain files of the same format (it contains: executable action scripts, environment files, and labels).

**Comment by vsoch** [comment]
Ah,this is true. Let's see...

`singularity.meta`
`singularity.info`
`singularity.db`
`singularity.setup`
`singularity.guts`
`singularity.brain`
`singularity.conf`
`singularity.dump`
`singularity.poop`
`singularity.make`
`singularity.uh`
`singularity.ah`
`singularity.oh`
`singularity.ar` (pirate!)
`singularity.booty`
`singularity.fish`
`singularity.owl`
`singularity.tree`
`singularity.awk`
`singularity.find`
`singularity.app`
`singularity.bum`
`singularity.bam`
`singularity.dig`
`singularity.egg` <--- ! like python! <3
   .green-egg
   .ham
   .sam-i-am
`singularity.gem` <-- like ruby?
`singularity.yo`
`singularity.mem`
`singularity.rom`
`singularity.um`
`singularity.nix`
`singularity.run`
`singularity.bag`
`singularity.hat`
     /cat <--- heh
`singularity.lol` <--- lol
`singularity.hi`
`singularity.hello`
`singularity.os`
`singularity.alt`
`singularity.data`
`singularity.head`
`singularity.store`
`singularity.stuff`
`singularity.greg`
`singularity.sel` (script, environ, labels)
`singularity.sem` (script,environ, meta)
`singularity.in`
`singularity.io`
`singularity.tray`
`singularity.browse`
`singularity.this`
`singularity.bin`
`singularity.etc`
`singularity.meta`
`singularity.pack`
`singularity.files`
`singularity.dino`
`singularity.state`
`singularity.star`
`singularity.greeneggsandhamsamIam`

lol I just laughed so hard I blew protein powder from a cup into my eyes. I think we're done here, I've demonstrated my prowess of 2-5 letter words on the 1st grade level.

**Comment by gmkurtzer** [comment]
HAHA!

What about the reasoning for not making it hidden? I'm not disagreeing, but I can think of reasons to hide it, so I'd like to hear some reasons to put it in plain view.

**Comment by satra** [comment]
@gmkurtzer - since i was the one who initiated the do not hide sentiment, here are my reasons:

1. hidden requires specialized knowledge, whereas visible makes people at least consider what the object is. most may ignore, but it's right in front of them.
2. this folder is exactly where i need to make most changes when i convert from docker to singularity (since all startup stuff goes in there).
3. in the past i would point to `/environment` and say this is like your `.bash*` (which i personally have never liked as a hidden file). i prefer that anything that directly impacts the running of my stuff should be very visible to me.

singularity can easily (i think) put this folder in `/etc` if the intent is to avoid clutter in `/`

may i ask the flip question as to why you wanted it to be hidden :) ?

**Comment by vsoch** [comment]
I can take mia culpa for pushing for the original folder to be hidden, and I'm on the fence now because I have mixed feelings!

1. I agree that the user should not have this information hidden, however editing our raw config files (when it may not be totally clear what is what, or what kinds of errors the changes could propagate)  is not the most user friendly workflow. For this first go, the folder serves as a metadata and all other things place dump of things (hence all the extension jokes alluding to that), which arguably, given we do this right (eventually) should never need to be manually picked through by the user. This is why my preference was for hidden. It should be discouraged to do these kinds of manual changes. 

However, hiding the folder only works given that we have provided easy (controlled) ways for the user to see the contents and make these changes. So far, we give "easy access" by way of sections in the bootstrap. That's pretty good, but it doesn't seem to allow for what Satra wants to do, which is more of an interactive use case.

In a nutshell, my ideal use case would be hidden folder, with tools / functions / etc. that make it "seeable" and editable in other ways.

2. Does the addition of a `.` significantly hinder your workflow here?

For me, hidden implies that it isn't explicitly part of the contents of the container. A lot of software, etc. skip over hidden files, and obviously they aren't viewable so a user would think it's part of the application itself. It's really trying to mirror a header, which again, one cannot see, but can interact with in other ways.

**Comment by vsoch** [comment]
Here is a more refined list of the ones that aren't just for funny :)

singularity.app
singularity.db
singularity.conf
singularity.egg
singularity.meta
singularity.nix
singularity.os
singularity.run
singularity.data
singularity.head
singularity.in
singularity.bin
singularity.etc
singularity.pack
singularity.files

**Comment by gmkurtzer** [comment]
@satra, You have some good points, I will address:

1. This is reasonable, but I'll throw out that the level of technical knowledge necessary to deal with hidden files in Unix is probably much lower then the barrier of knowledge necessary to understand containers. ;)
2. I would argue that nothing in this directory should be edited. It gets created at bootstrap/import time, and people should be using the definition file to populate these files.
3. You bring up a great point, but I'm gonna swipe it and use it to articulate my point! ;)  The `.bash*` files are hidden as well, and for the same reasons of the generally accepted file system nomenclature and pedantics on hidden login files I would propose that the singularity folder should also be hidden.

You have a point about putting it in `/etc` and I thought of that too. But because of the executable action scripts which are not designed to be configurable (`/etc` is supposed to be reserved for configuration files as defined by the Linux FSH) I do not consider `/etc` an appropriate location. 

I think the above articulates why it should be hidden, but to summarize... It should not be generally edited directly but relevant user facing bits (e.g. `/environment`) are indeed visible via symlinks (something I'm not also very fond of, but it has precedent). Additionally, keeping it hidden also keeps it just outside of the visibility and general usage of the encapsulated container; which is exactly what we want. Additionally, because everything in this directory is for the purpose of Singularity alone, keeping it orthogonal to contained system also makes sense.

Hope that helps to elaborate on what I'm thinking.

**Comment by vsoch** [comment]
I think we are good on this now, folder is called `/.singularity.d` closing!
